### PR TITLE
add coverage for removing all user settings on deletion + test coverage

### LINKED ISSUE #23672: [Bug] Deleted user’s settings persist when username is reused
### What happened?

**Description**

After a user has been deleted from a Matomo instance, their personal settings is retained when adding another user with the same username (different email).

### What should happen?

**Expected and Actual results:**

The deleted user's username should not be made available again. Instead currently settings are retained:

Image

### How can this be reproduced?

**Steps to reproduce**
1. Create a user, on a Matomo instance, called victimUser
2. Log in as victimUser and change personal settings randomly and keep a note of the changes
3. Delete the user
4. Create a second user with username victimUser (different email this time)
5. Log in as victimUser and compare personal settings with previous changes

### Matomo version

latest

### PHP version

_No response_

### Server operating system

_No response_

### What browsers are you seeing the problem on?

_No response_

### Computer operating system

_No response_

### Relevant log output

```shell

```

### Validations

- [x] Read our [Contributing Guidelines](https://github.com/matomo-org/matomo/blob/5.x-dev/CONTRIBUTING.md).
- [x] Follow our [Security Policy](https://github.com/matomo-org/matomo/blob/5.x-dev/SECURITY.md).
- [x] Check that there isn't already an issue that reports the same bug to avoid creating duplicates.
- [x] The provided steps to reproduce is a [minimal reproducible](https://stackoverflow.com/help/minimal-reproducible-example) of the Bug.

### PR DESCRIPTION
**Pull request by nathangavin**
### Description

Please include a description of this change and which issue it fixes. If no issue exists yet please include context and what problem it solves.

fixes #23672

### Checklist

- [✔] I have understood, reviewed, and tested all AI outputs before use
- [✔] All AI instructions respect security, IP, and privacy rules

### Review

- [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
- [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behaviour of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
- [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
- [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
- [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
- [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
- [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
- [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
- [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
- [ ] [New documentation added or existing documentation updated](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)

### CODE REVIEW

**Review by chippison — COMMENTED** [review]
I've checked the code and done some functional tests, overall it everything works as intended. Deleting a user does remove the settings from the DB

There are a couple of edge case issues codex found though:

[P1] User deletion can leave legacy per-user data behind when one observer throws
Basically, if a plugin throws an exception when deleting a user, it could leave some data behind and not clean it up 
 - Was able to recreate this by creating a local plugin that throws exception on UserDelete
 - Codex recommendations (you can just use this to get a starting point): 
    1. Add direct legacy cleanup in deleteUserOptions() for MobileMessaging + ProfessionalServices keys (smallest blast radius).
   2. Keep event-based cleanup, but isolate observer failures per callback for this event path.


[P2] Migration strategy for ProfessionalServices scales poorly with user count
 - wasn't really able to recreate this one, but seems like some DB query optimisation needed 
 - Codex recommendations: (you can just use this to get a starting point): 
    1. Load ProfessionalServices.DismissedWidget.% once, group by login in PHP, then write/delete once.
   2. Keep logic but batch by login chunks with tighter SQL (still more churn than option 1).

File: plugins/UsersManager/Model.php [hunk 1]
```diff
@@ -787,12 +790,36 @@ public function deleteUserOnly($userLogin)
...
          * @param string $userLogins The login handle of the deleted user.
          */
-        Piwik::postEvent('UsersManager.deleteUser', array($userLogin));
+        try {
+            Piwik::postEvent('UsersManager.deleteUser', array($userLogin));
+        } catch (\Throwable $e) {
+            error_log(sprintf(
+                'Error while processing event UsersManager.deleteUser for login "%s": %s',
+                $userLogin,
+                $e->getMessage()
```
Comment (chippison) [inline plugins/UsersManager/Model.php]: Are we good to log user data directly in the logs? Wouldn't this risk user data leak? 

Also instead of using raw `error_log`
```suggestion
            StaticContainer::get(LoggerInterface::class)->error('Something failed', [
    'exception' => $e,
]);
```

File: plugins/UsersManager/Model.php [hunk 2]
```diff
@@ -787,12 +790,36 @@ public function deleteUserOnly($userLogin)
...
+        Option::deleteLike('ProfessionalServices.DismissedWidget.%.' . $userLogin);
+
+        $preferences = [
+            API::PREFERENCE_DEFAULT_REPORT,
+            API::PREFERENCE_DEFAULT_REPORT_DATE,
+            'isLDAPUser',
+            'hideSegmentDefinitionChangeMessage',
+        ];
+        $customPreferences = StaticContainer::get('usersmanager.user_preference_names');
+        $preferences = array_merge($preferences, $customPreferences);
```
Comment (chippison) [inline plugins/UsersManager/Model.php:824]: `array_merge($preferences, $customPreferences)` assumes `usersmanager.user_preference_names` is always an array. If a plugin returns a scalar, this is a hard TypeError in PHP 8+, and user deletion breaks.


```suggestion
       // From LegacyUserSettingsMigration
        if (!is_array($customPreferenceNames)) {
            $customPreferenceNames = [];
        }
        $customPreferenceNames = array_values(array_filter($customPreferenceNames, 'is_string'));
```

File: plugins/UsersManager/Model.php [hunk 3]
```diff
@@ -787,12 +790,36 @@ public function deleteUserOnly($userLogin)
          *
          * @param string $userLogins The login handle of the deleted user.
          */
-        Piwik::postEvent('UsersManager.deleteUser', array($userLogin));
+        try {
```
Comment (chippison) [inline plugins/UsersManager/Model.php:794]: Codex found another issue for this: 
If `Piwik::postEvent('UsersManager.deleteUser', array($userLogin));` fails, there is a chance that there are some values that will become stale for UsersManager plugin, but I guess this is why we log the error so we will know which fails  
As long as we are  aware of this and are ok with the possibility of having some stale data when this errors out, I think I'm good with this.

Comment (nathangavin) [reply plugins/UsersManager/Model.php:794]: Yeah based on your earlier suggestion around perhaps handling exceptions thrown for each event handled in a postEvent, I decided that it was better to handle any exceptions thrown as a whole outside the event handling, and risk the data not being removed. I believe this method of exception handling with events is how we handle it in other places.

File: plugins/UsersManager/Model.php (hunk 1 above)
Comment (nathangavin) [reply plugins/UsersManager/Model.php]: yeah good point, used the proper class and removed the username from the message. It may still appear in the exception message though.

File: plugins/UsersManager/Model.php (hunk 2 above)
Comment (nathangavin) [reply plugins/UsersManager/Model.php:824]: agreed, fixed.

**Comment by chippison** [comment]
I have checked the changes and it looks good from my end. 
I also have done another round of testing and can confirm that:
 - when a user is deleted, its legacy settings are also deleted. 
 - when a new user is created using the previous username, it does not follow/reuse the settings of the old user.

**Review by chippison — DISMISSED** [review]

File: core/Updates/5.8.0-b2.php [hunk 4]
```diff
@@ -29,6 +31,10 @@ public function getMigrations(Updater $updater)
     {
         return [
             $this->migration->db->addColumn('archiving_metrics', 'is_temporary', 'TINYINT(1) UNSIGNED NOT NULL DEFAULT 0'),
+            new CustomMigration(
```
Comment (sgiehl) [inline core/Updates/5.8.0-b2.php]: Please always create a new migration script if migrations are required. Never add something to a migration script that already exists, as that risks migrations not to be executed safely.

File: plugins/ProfessionalServices/ProfessionalServices.php [hunk 5]
```diff
@@ -35,10 +36,16 @@ public function registerEvents()
...
             'Template.afterPagePerformanceReport' => 'getSeoWebVitalsPromo',
             'Template.afterSearchEngines' => 'getSeoWebVitalsPromo',
+            'UsersManager.deleteUser' => 'onDeleteUser',
             'Translate.getClientSideTranslationKeys' => 'getClientSideTranslationKeys',
         );
     }
 
+    public function onDeleteUser($userLogin): void
+    {
+        Option::deleteLike('ProfessionalServices.DismissedWidget.%.' . $userLogin);
```
Comment (sgiehl) [inline plugins/ProfessionalServices/ProfessionalServices.php]: Shouldn't that happen in the migration and afterwards be unneeded?

File: core/Settings/Storage/LegacyUserSettingsMigration.php [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        $prefix = 'ProfessionalServices.DismissedWidget.';
+        $legacySettings = Option::getLike($prefix . '%');
+        $legacyByLogin = [];
+
+        foreach ($legacySettings as $optionName => $optionValue) {
+            if (!str_starts_with($optionName, $prefix)) {
+                continue;
+            }
+
+            $lastSeparatorPosition = strrpos($optionName, '.');
```
Comment (sgiehl) [inline core/Settings/Storage/LegacyUserSettingsMigration.php]: This will not behave correctly if a login contains a `.`.

File: core/Settings/Storage/LegacyUserSettingsMigration.php [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+                if ($login === '' || !$this->isLikelyValidLogin($login)) {
+                    continue;
+                }
+
+                if (isset($knownLogins[$login]) && !$this->hasSetting('UsersManager', $login, $preferenceName)) {
+                    $this->store->set('UsersManager', $login, $preferenceName, $optionValue);
+                    ++$migratedCount;
+                }
+
+                if ($preferenceName === 'isLDAPUser' && isset($knownLogins[$login])) {
```
Comment (sgiehl) [inline core/Settings/Storage/LegacyUserSettingsMigration.php:220]: I haven't checked LoginLdap: Does it write and read the Option directly or why does it need a special handling?

File: core/Settings/Storage/UserScopedSettingsStore.php [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        $this->factory = $factory;
+    }
+
+    /**
+     * @param mixed $defaultValue
+     * @return mixed
+     */
+    public function get(string $pluginName, string $userLogin, string $key, $defaultValue = null)
+    {
+        $settings = $this->getAll($pluginName, $userLogin);
```
Comment (sgiehl) [inline core/Settings/Storage/UserScopedSettingsStore.php]: I think the handling in this class looks a bit prone to race conditions. All methods are first reading all values, then manipulating some of them in order to write everything again.
If that happens in two sessions at the same time it may happen that parallel requests might overwrite each other by accident.
If we refactor this to directly use the PluginSettingsTable, rather then the Factory, we can easily extend the Store to have read and write operations for single entries.

File: core/Settings/Storage/UserScopedSettingsStore.php [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+ * @link    https://matomo.org
+ * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\Settings\Storage;
+
+use Piwik\Common;
+use Piwik\Db;
+
+class UserScopedSettingsStore
```
Comment (sgiehl) [inline core/Settings/Storage/UserScopedSettingsStore.php]: I'm actually not fully happy with the naming of this class. Technically seen it not really a Store, as it does not store anything itself. It's more like a Facade or Repository that provides access to the already existing PluginSettingsTable Store. 
It might furthermore be better not to mix up things. The code below uses the existing Store for almost everything, but then has a direct SQL access to fetch all data. I think the scope should be cut more clearly. So this class would be sort of an access layer for the store, and the store itself should then hold all methods for accessing data.

File: plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js [hunk 10]
```diff
@@ -6,6 +6,32 @@
  * @link    https://matomo.org
  * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
  */
+describe("MobileMessaging", function () {
```
Comment (sgiehl) [inline plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js:9]: Any specific reason for moving this one into it's own block?

File: plugins/MobileMessaging/MobileMessaging.php [hunk 11]
```diff
@@ -90,6 +91,11 @@ public function getStylesheetFiles(&$stylesheets)
         $stylesheets[] = "plugins/MobileMessaging/stylesheets/MobileMessagingSettings.less";
     }
 
+    public function onDeleteUser($userLogin): void
+    {
+        Option::delete($userLogin . self::USER_SETTINGS_POSTFIX_OPTION);
```
Comment (sgiehl) [inline plugins/MobileMessaging/MobileMessaging.php]: Might also be unneeded if the migration already cleaned things up.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Didn't have time for any testing and only performed a rough code review

File: core/Updates/5.8.0-b2.php (hunk 4 above)
Comment (nathangavin) [reply core/Updates/5.8.0-b2.php]: Moved to a new migration file, updated version appropriately.

File: plugins/ProfessionalServices/ProfessionalServices.php (hunk 5 above)
Comment (nathangavin) [reply plugins/ProfessionalServices/ProfessionalServices.php]: Yep, removed.

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 6 above)
Comment (nathangavin) [reply core/Settings/Storage/LegacyUserSettingsMigration.php]: replaced this to use knownlogins to work out the login and plugin name.

File: core/Settings/Storage/UserScopedSettingsStore.php (hunk 8 above)
Comment (nathangavin) [reply core/Settings/Storage/UserScopedSettingsStore.php]: rewrote this and the pluginsSettingsTable class to move all db touching points to the storage class, this class is now a wrapper.

File: core/Settings/Storage/UserScopedSettingsStore.php (hunk 9 above)
Comment (nathangavin) [reply core/Settings/Storage/UserScopedSettingsStore.php]: see other comment, also renamed class to make more sense.

File: plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js (hunk 10 above)
Comment (nathangavin) [reply plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js:9]: This is the easiest way to manage the state of the tests, clearing the PluginsSettingTable in the test env would require an update to the related repo, this is simpler.

File: plugins/MobileMessaging/MobileMessaging.php (hunk 11 above)
Comment (nathangavin) [reply plugins/MobileMessaging/MobileMessaging.php]: removed.

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 7 above)
Comment (nathangavin) [reply core/Settings/Storage/LegacyUserSettingsMigration.php:220]: `plugins/UsersManager/API::setUserPreference()` sets a bunch of options, including `isLDAPUser`. This keeps using the old method so that LoginLDAP doesn't require an update to work with this bugfix.

File: core/Settings/Storage/Backend/PluginSettingsTable.php [hunk 12]
```diff
@@ -201,4 +344,58 @@ public static function removeAllSettingsForPlugin($pluginName)
             }
         }
     }
+
+    private static function getColumns(): array
```
Comment (sgiehl) [inline core/Settings/Storage/Backend/PluginSettingsTable.php:351]:
```suggestion
    /**
     * @return string[]
     */
    private static function getColumns(): array
```

File: core/Settings/Storage/Backend/PluginSettingsTable.php [hunk 13]
```diff
@@ -96,6 +96,104 @@ public function save($values)
...
+     * Save exactly one setting key atomically for the current plugin/user context
+     *
+     * if the value is null, the key is deleted.
+     *
+     * @throws Exception when param $settingName is empty
+     */
+    /**
+     * @param mixed $value
+     */
+    public function saveValue(string $settingName, $value): void
```
Comment (sgiehl) [inline core/Settings/Storage/Backend/PluginSettingsTable.php:109]: Some tests for all the new methods could be helpful

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 7 above)
Comment (sgiehl) [reply core/Settings/Storage/LegacyUserSettingsMigration.php:220]: would be helpful to create a PR or at least an issue for LoginLdap already to ensure future versions of the plugin will no longer rely an the option for being set. In addition we should add a `@deprecated` comment here and in API, mentioning that it should be removed with Matomo 6.

File: core/Settings/Storage/LegacyUserSettingsMigration.php [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+                    // Keep legacy option key for LoginLdap submodule compatibility.
+                    Option::set($optionName, $optionValue);
```
Comment (sgiehl) [inline core/Settings/Storage/LegacyUserSettingsMigration.php]: Is this actually needed. We are trying to convert options to something else right? So the option here should already exist and simply skipping the `delete` below should already be enough.

File: core/Updates/5.9.0-b1.php [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+use Piwik\Updates;
+
+class Updates_5_9_0_b1 extends Updates
+{
+    public function getMigrations(Updater $updater)
+    {
+        return [
+            new CustomMigration(
+                [MigrateUserScopedSettings::class, 'migrate'],
+                './console core:matomo590Repository-migrate-user-scoped-settings'
```
Comment (sgiehl) [inline core/Updates/5.9.0-b1.php]:
```suggestion
                './console core:matomo590-migrate-user-scoped-settings'
```

File: plugins/MobileMessaging/Model.php [hunk 16]
```diff
@@ -264,27 +348,70 @@ private function getPhoneNumbersFromSettings(string $login): array
...
+        }
+
+        return [
+            'verified' => !empty($verificationData['verified']),
+            'verificationCode' => isset($verificationData['verificationCode']) && $verificationData['verificationCode'] !== null
+                ? (string) $verificationData['verificationCode']
+                : null,
+            'verificationTime' => isset($verificationData['verificationTime']) && $verificationData['verificationTime'] !== null
+                ? (int) $verificationData['verificationTime']
+                : null,
```
Comment (sgiehl) [inline plugins/MobileMessaging/Model.php]: Isset already ensures that a value is not `null`, so the checks are redundant.
```suggestion
            'verificationCode' => isset($verificationData['verificationCode'])
                ? (string) $verificationData['verificationCode']
                : null,
            'verificationTime' => isset($verificationData['verificationTime'])
                ? (int) $verificationData['verificationTime']
                : null,
```

File: plugins/ProfessionalServices/PromoWidgetDismissal.php [hunk 17]
```diff
@@ -34,11 +42,27 @@ public function isPromoWidgetDismissedForCurrentUser(string $widgetName): bool
...
+        }
+
+        return [];
     }
 
-    private function getDismissedWidgetOptionName(string $widgetName): string
+    private function getAccessManager(): UserScopedSettingsAccessManager
     {
-        return sprintf(self::DISMISSED_WIDGET_OPTION_NAME, $widgetName, Piwik::getCurrentUserLogin());
+        return StaticContainer::get(UserScopedSettingsAccessManager::class);
```
Comment (sgiehl) [inline plugins/ProfessionalServices/PromoWidgetDismissal.php]: As far as I have seen it should be possible to change that one to use DI in a constructor instead. Not necessarily needed, but would be more robust and future proof.

File: plugins/UsersManager/API.php [hunk 18]
```diff
@@ -224,8 +225,13 @@ public function setUserPreference($userLogin, $preferenceName, $preferenceValue)
             Piwik::checkUserHasSuperUserAccess();
         }
 
-        $nameIfSupported = $this->getPreferenceId($userLogin, $preferenceName);
-        Option::set($nameIfSupported, $preferenceValue);
+        $this->assertPreferenceNameIsSupported($preferenceName);
+        $this->getUserSettingsAccessManager()->set('UsersManager', $userLogin, $preferenceName, $preferenceValue);
+
+        // Keep legacy option key for LoginLdap compatibility without requiring submodule changes.
```
Comment (sgiehl) [inline plugins/UsersManager/API.php]:
```suggestion
        // Keep legacy option key for compatibility with older LoginLdap versions.
```

File: plugins/UsersManager/Model.php [hunk 19]
```diff
@@ -787,12 +791,40 @@ public function deleteUserOnly($userLogin)
...
+        if (empty($customPreferences)) {
+            $customPreferences = [];
+        } elseif (!is_array($customPreferences)) {
+            $customPreferences = [$customPreferences];
+        }
+        $preferences = array_merge($preferences, $customPreferences);
+
+        foreach ($preferences as $preference) {
+            Option::delete($userLogin . API::OPTION_NAME_PREFERENCE_SEPARATOR . $preference);
+        }
```
Comment (sgiehl) [inline plugins/UsersManager/Model.php:829]: Those parts, which are removing legacy option, that should not longer be used/added, could maybe be marked as `// @todo  remove with Matomo 6` so we do not keep them for forever.

File: plugins/UsersManager/Model.php (hunk 3 above)
Comment (sgiehl) [reply plugins/UsersManager/Model.php:794]: I guess it's a more general problem. Methods called during an event handling should actually not throw an exception at all, but handle everything correctly. If that would be the case it would be fine to just trigger the vent.
The change should be fine for now, but I don't think we should start doing that everywhere, but instead fix the event handler, to handle exceptions correctly.

File: plugins/CoreAdminHome/tests/Integration/Commands/MigrateUserScopedSettingsTest.php [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+ * @group CoreAdminHome
+ * @group CoreAdminHome_Integration
+ */
+class MigrateUserScopedSettingsTest extends IntegrationTestCase
+{
+    public function testMigrateMovesKnownUsersToPluginSettingsAndPurgesLegacyOptions(): void
+    {
+        $usersModel = new UsersModel();
+        $store = new UserScopedSettingsAccessManager();
+        $activeUser = 'activeMigrUser';
```
Comment (sgiehl) [inline plugins/CoreAdminHome/tests/Integration/Commands/MigrateUserScopedSettingsTest.php:30]: Could be useful to also test that login containing a `.` work as expected.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Tried doing a more detailed review now and left some further comments around possible improvements/fixes.

File: core/Settings/Storage/Backend/PluginSettingsTable.php (hunk 13 above)
Comment (nathangavin) [reply core/Settings/Storage/Backend/PluginSettingsTable.php:109]: Added

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 14 above)
Comment (nathangavin) [reply core/Settings/Storage/LegacyUserSettingsMigration.php]: yep agreed, skipping the delete is sufficient.

File: plugins/UsersManager/Model.php (hunk 19 above)
Comment (nathangavin) [reply plugins/UsersManager/Model.php:829]: added various @deprecated and @todo where appropriate

File: core/Settings/Storage/Backend/PluginSettingsTable.php [hunk 21]
```diff
@@ -201,4 +344,61 @@ public static function removeAllSettingsForPlugin($pluginName)
             }
         }
     }
+
+    /**
+     * return string[]
```
Comment (sgiehl) [inline core/Settings/Storage/Backend/PluginSettingsTable.php]:
```suggestion
     * @return string[]
```

File: core/Settings/Storage/LegacyUserSettingsMigration.php [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+use Piwik\Db;
+use Piwik\Option;
+use Piwik\Piwik;
+use Piwik\Plugins\MobileMessaging\MobileMessaging;
+use Piwik\Plugins\UsersManager\API as UsersManagerAPI;
+
+/**
+ * @deprecated Remove this class with Matomo 6 once legacy user-scoped option migration is no longer needed.
+ * @todo Remove this class and its callers with Matomo 6.
+ */
```
Comment (sgiehl) [inline core/Settings/Storage/LegacyUserSettingsMigration.php]: We actually can't remove that, as people might directly update to Matomo 6, while skipping any version that would have done the migration.

```suggestion
```

File: plugins/CoreAdminHome/Commands/MigrateUserScopedSettings.php [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+namespace Piwik\Plugins\CoreAdminHome\Commands;
+
+use Piwik\Container\StaticContainer;
+use Piwik\Plugin\ConsoleCommand;
+use Piwik\Settings\Storage\LegacyUserSettingsMigration;
+
+/**
+ * @deprecated Remove this command with Matomo 6 after legacy user-scoped option migration is no longer required.
+ * @todo Remove this command with Matomo 6.
+ */
```
Comment (sgiehl) [inline plugins/CoreAdminHome/Commands/MigrateUserScopedSettings.php]:
```suggestion
```

File: plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js (hunk 10 above)
Comment (sgiehl) [reply plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js:9]: So I guess that now only works because there is still the legacy option handling in place. That way the test will start to fail, once we remove the legacy handling. Should be fine for now, but in general it might be better to update tests directly so they are not supposed to fail at some point. For that one, guess a custom Fixture would have done the trick.

File: core/Settings/Storage/LegacyUserSettingsMigration.php [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+            $legacySettings = Option::getLike('%' . $suffix);
+
+            foreach ($legacySettings as $optionName => $optionValue) {
+                if (!str_ends_with($optionName, $suffix)) {
+                    continue;
+                }
+
+                $login = substr($optionName, 0, -strlen($suffix));
+                if ($login === '' || !$this->isLikelyValidLogin($login)) {
+                    continue;
```
Comment (sgiehl) [inline core/Settings/Storage/LegacyUserSettingsMigration.php:208]: This most likely should not occur. But in case it does - should we remove the Option nevertheless? Or could it be the case that we then remove something that should remain?

**Review by sgiehl — CHANGES_REQUESTED** [review]
some last comments, afterwards that should be good to go.

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 24 above)
Comment (nathangavin) [reply core/Settings/Storage/LegacyUserSettingsMigration.php:208]: Looking into it a bit more, looks like the types of options that would be caught by this check (`_<pref>` or `<invalidLogin>_<pref>`) are very unlikely to exist anywhere, certainly none exist at the moment. Though looking into it more, looks like it would be possible to use a custom plugin that registers a preference key, like 'core', as well as a user with a 'protected' username, like 'version', to remove some options like 'version_core'. I don't really think we can handle that scenario though.

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 22 above)
Comment (nathangavin) [reply core/Settings/Storage/LegacyUserSettingsMigration.php]: right of course, that makes sense. Removed.

File: plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js (hunk 10 above)
Comment (nathangavin) [reply plugins/MobileMessaging/tests/UI/MobileMessaging_spec.js:9]: yea ok good point about using a new fixture, I've implemented that approach so it shouldn't fail going forward now.

**Review by sgiehl — APPROVED** [review]

File: core/Settings/Storage/LegacyUserSettingsMigration.php (hunk 24 above)
Comment (sgiehl) [reply core/Settings/Storage/LegacyUserSettingsMigration.php:208]: Let's just ignore that for now then.
